From d8e0927334b7c4f7823aba278993d6bc399881cc Mon Sep 17 00:00:00 2001
From: "Andrew D. Gazizov" <agazizov@qti.qualcomm.com>
Date: Thu, 6 Feb 2025 16:44:39 +0100
Subject: [PATCH] virgl: Implement API to attach dmabuf backing for classic 3d
 res.

It necessary to be able import classic 3d resources to Venus in same
way as it happens for blob resources and allowing provide support of
VK_ANDROID_external_memory_android_hardware_buffer for Android guests
in order to import into Vulkan AHB with 3d classic resource backing.

New API functions:
VIRGL_EXPORT int virgl_renderer_resource_attach_dmabuf(int res_handle, int fd);
VIRGL_EXPORT void virgl_renderer_resource_detach_dmabuf(int res_handle);

Signed-off-by: "Andrew D. Gazizov" <agazizov@qti.qualcomm.com>
---
 src/virglrenderer.c | 31 +++++++++++++++++++++++++++++++
 src/virglrenderer.h |  3 +++
 2 files changed, 34 insertions(+)

diff --git a/src/virglrenderer.c b/src/virglrenderer.c
index ba886400..65b5de24 100644
--- a/src/virglrenderer.c
+++ b/src/virglrenderer.c
@@ -410,6 +410,37 @@ void virgl_renderer_resource_detach_iov(int res_handle, struct iovec **iov_p, in
    virgl_resource_detach_iov(res);
 }
 
+int virgl_renderer_resource_attach_dmabuf(int res_handle, int fd)
+{
+   TRACE_FUNC();
+   struct virgl_resource *res = virgl_resource_lookup(res_handle);
+   if (!res)
+      return EINVAL;
+
+   if (res->fd != -1 || res->fd_type != VIRGL_RESOURCE_FD_INVALID)
+      return EINVAL;
+
+   /* Setting fd for Venus import without ownership transfer */
+   res->fd = fd;
+   res->fd_type = VIRGL_RESOURCE_FD_DMABUF;
+   res->map_size = lseek(fd, 0, SEEK_END);
+
+   return 0;
+}
+
+void virgl_renderer_resource_detach_dmabuf(int res_handle)
+{
+   TRACE_FUNC();
+   struct virgl_resource *res = virgl_resource_lookup(res_handle);
+   if (!res)
+      return;
+
+   /* Ownership remains on client side so just unset fd */
+   res->fd = -1;
+   res->fd_type = VIRGL_RESOURCE_FD_INVALID;
+   res->map_size = 0;
+}
+
 int virgl_renderer_create_fence(int client_fence_id, UNUSED uint32_t ctx_id)
 {
    TRACE_FUNC();
diff --git a/src/virglrenderer.h b/src/virglrenderer.h
index 4c8682e0..e068837a 100644
--- a/src/virglrenderer.h
+++ b/src/virglrenderer.h
@@ -332,6 +332,9 @@ VIRGL_EXPORT int virgl_renderer_resource_attach_iov(int res_handle, struct iovec
                                                     int num_iovs);
 VIRGL_EXPORT void virgl_renderer_resource_detach_iov(int res_handle, struct iovec **iov, int *num_iovs);
 
+VIRGL_EXPORT int virgl_renderer_resource_attach_dmabuf(int res_handle, int fd);
+VIRGL_EXPORT void virgl_renderer_resource_detach_dmabuf(int res_handle);
+
 VIRGL_EXPORT int virgl_renderer_create_fence(int client_fence_id, uint32_t ctx_id);
 
 VIRGL_EXPORT void virgl_renderer_force_ctx_0(void);
-- 
2.34.1

